### DISCUSSION TITLE
Trouble implementing fallback to different service

### DISCUSSION THREAD

**Discussion by a082651** [App-vNext/Polly#1973] [Q&A] [answered]
I'm trying to follow the guidance in the [fallback documentation](https://www.pollydocs.org/strategies/fallback.html#using-retry-for-fallback), specifically under the anti-pattern "Using retry for fallback". The code for using a Fallback strategy doesn't work as written because it lacks a `FallbackAction`. This seems to conflict with the listed benefit of "eliminating the need for additional code like Context or ExecuteOutcomeAsync()", as `FallbackAction` seems to require interaction with one of those.

We are rolling out a new API that replaces an old one. I'd like to use the old one as a fallback to reduce the impact of our rollout. I can't seem to find an implementation of `FallbackAction` that has access to the value returned by the `OnFallback` callback.

Here is a highly simplified example:

```
async Task<string> FallbackStrategy() => "fallback";
async Task<string> FlaggedStrategy() => "flagged";

var resiliencePipeline = new ResiliencePipelineBuilder<string>()
    .AddFallback(new FallbackStrategyOptions<string>
    {
        ShouldHandle = new PredicateBuilder<string>().HandleResult(_ => true),
        OnFallback = async x => await FallbackStrategy(),
        FallbackAction = x => Outcome.FromResultAsValueTask<string>("outcome"),
    })
    .Build();

string actual = await resiliencePipeline.ExecuteAsync(async _ => await FlaggedStrategy());
Assert.Equal("fallback", actual); // fails, actual = "outcome"
```

**Comment by peter-csala** [comment, marked as answer]
Oops, that's my bad. Let me file a [PR](https://github.com/App-vNext/Polly/pull/1976) today to fix that sample first.

After that I will take a look at your other problem.

**Reply by a082651** [reply]
Changing FallbackAction to `FallbackAction = async args => Outcome.FromResult(await FallbackStrategy())` has the expected result. Thank you for the quick turnaround!

**Comment by peter-csala** [comment]
> I can't seem to find an implementation of `FallbackAction` that has access to the value returned by the `OnFallback` callback.

It works in the other way around:

1.  The `FallbackAction` is executed 

https://github.com/App-vNext/Polly/blob/de7fdb042aa91a1bb3e28050ee978d380cc597b2/src/Polly.Core/Fallback/FallbackResilienceStrategy.cs#L22

2. The `OnFallback` args is created 

https://github.com/App-vNext/Polly/blob/de7fdb042aa91a1bb3e28050ee978d380cc597b2/src/Polly.Core/Fallback/FallbackResilienceStrategy.cs#L29

3. The `OnFallback` is invoked

https://github.com/App-vNext/Polly/blob/de7fdb042aa91a1bb3e28050ee978d380cc597b2/src/Polly.Core/Fallback/FallbackResilienceStrategy.cs#L35

So, you can access the result of the `FallbackAction`'s result inside the `OnFallback` via the `args` parameter's [`Outcome`](https://github.com/App-vNext/Polly/blob/main/src/Polly.Core/Fallback/OnFallbackArguments.cs#L28) property.
